src/documentation/programming-guide.md

```diff
@@ -900,8 +901,190 @@ odds = results.odd
 tens = results[None]  # the undeclared main output
 ```
 
-<a name="io"></a>
-<a name="running"></a>
+## <a name="io"></a>Pipeline I/O
+
+When you create a pipeline, you often need to read data from some external source, such as a file in cloud storage or a database. Likewise, you may want your pipeline to output its result data to a similar external data sink. Beam provides read and write transforms for a number of common data storage types. If you want your pipeline to read from or write to a data storage format that isn't supported by the built-in transforms, you can implement your own read and write transforms.
+
+> A guide that covers how to implement your own Beam IO transforms is in progress ([BEAM-1025](https://issues.apache.org/jira/browse/BEAM-1025)).
+
+### Reading Input Data
+
+Read transforms read data from an external source and return a `PCollection` representation of the data for use by your pipeline. You can use a read transform at any point while constructing your pipeline to create a new `PCollection`, though it will be most common at the start of your program. Here are examples of two common ways to read data.
+
+#### Reading from a `Source`:
+
+```java
+// This example uses XmlSource.
+XmlSource<String> source =   XmlSource.<String>from(file.toPath().toString())
+     .withRootElement("root")
+     .withRecordElement("record")
+     .withRecordClass(Record.class);
+PCollection<String> output = pipeline.apply(Read.from(source));
+```
+
+```python
+pipeline | beam.io.Read(beam.io.TextFileSource("protocol://path/to/some/inputData.txt"))
```

Comment:
For Python please use: pipeline | beam.io.ReadFromText('protocol://path/to/some/inputData.txt')
Current snippent uses a Dataflow native source which will be deprecated/removed soon. Source object itself is private so you will have to use the PTransform (ReadFromText) in the example.

As Davor mentioned, probably Java snippet should be changed to use the corresponding Read PTransform for Text (TextIO.Read).